// @rxweb/reactive-form-validators/esm5/directives/template-validations/async-validation.directive.js
import * as tslib_1 from "tslib";
import { Input, Directive, forwardRef } from '@angular/core';
import { NG_ASYNC_VALIDATORS } from '@angular/forms';
var AsyncValidationDirective = /** @class */ (function () {
    function AsyncValidationDirective() {
    }
    AsyncValidationDirective_1 = AsyncValidationDirective;
    AsyncValidationDirective.prototype.validate = function (control) {
        if (this.async)
            return this.async(control);
        return new Promise(function (resolve, reject) { resolve(null); });
    };
    var AsyncValidationDirective_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], AsyncValidationDirective.prototype, "async", void 0);
    AsyncValidationDirective = AsyncValidationDirective_1 = tslib_1.__decorate([
        Directive({
            selector: '[ngModel],[formControlName],[formControl]',
            providers: [{
                    provide: NG_ASYNC_VALIDATORS,
                    useExisting: forwardRef(function () { return AsyncValidationDirective_1; }),
                    multi: true
                }]
        })
    ], AsyncValidationDirective);
    return AsyncValidationDirective;
}());
export { AsyncValidationDirective };
